samza-api/src/main/java/org/apache/samza/startpoint/Startpoint.java

```diff
@@ -50,12 +50,13 @@ public long getCreationTimestamp() {
   }
 
   /**
-   * Apply the visitor {@link StartpointVisitor}'s register methods to the instance of this {@link Startpoint}
-   * class.
-   * @param systemStreamPartition The {@link SystemStreamPartition} needed to register with the {@link StartpointVisitor}
-   * @param startpointVisitor The visitor to register with.
+   * Applies the {@link StartpointVisitor}'s visit methods to the {@link Startpoint}
+   * and resolves it to a system specific offset.
+   * @param systemStreamPartition the {@link SystemStreamPartition} of the startpoint.
+   * @param startpointVisitor the visitor of the startpoint.
+   * @return the resolved offset.
    */
-  public abstract void apply(SystemStreamPartition systemStreamPartition, StartpointVisitor startpointVisitor);
+  public abstract String apply(SystemStreamPartition systemStreamPartition, StartpointVisitor startpointVisitor);
```

Comment:
One of the core concepts of the visitor pattern is to make the visitor generic. The caller is always going to use a specific implementation of the visitor, so they do need to know which one they are using.

Reply:
To clarify further, when the caller uses a visitor, the caller is directly using the concrete implementation of the visitor. So that usage is what defines the functional contract and return type. The base `StartpointVisitor` does not need to define the functional contract and return type.

Reply:
I think for now, generifying the inputs and outputs won't benefit much, but rather will make code readability a little harder. If anything, the output should be explicit as possible or else the interpretation is left to the caller. Also I don't see the input being something other than `SystemStreamPartition` in the future. 

I believe the major motivation for this API change is that other components of Samza that deal with offsets understand string/specific offsets only. IMO, if we want to generify this API, we should take a holistic approach and do it across all offset related components. It'll probably take more careful thought, so could we revisit this approach later?

Reply:
Discussed offline with @cameronlee314 @dnishimura .  Though there was some agreements on the points raised against generification in above comments, overall generifying the StartpointVisitor will allow disjoint processing control-flows to be integrated with startpoints in the future. We'd agreed to generify the `StartpointVisitor` contract.